tsdb/head_wal.go

```diff
@@ -469,6 +469,17 @@ Outer:
 	return nil
 }
 
+// getWALReplayConcurrency returns the maximum number of CPUs that can simultaneously processes WAL replay.
+// This value is taken from HeadOptions.WALReplayConcurrency, if it is positive and less than the number of available
+// CPUs that can be executing simultaneously. Otherwise, GOMAXPROCS is returned.
+func (h *Head) getWALReplayConcurrency() int {
```

Comment:
_This comment is about keeping a design closer to the other Head options._

I'm wondering if this logic should be moved to `DefaultHeadOptions()`, where we just initialise `WALReplayConcurrency` to `GOMAXPROCS`, and then the WAL replay logic will simply honor `h.opts.WALReplayConcurrency`, whatever value it is. If you want to do any validation, then it should be done in `NewHead()` (e.g. see validation of `opts.ChunkRange` as a reference).

Reply:
+1 as well. We can remove this function, update `DefaultHeadOptions()`, and do the validation in `NewHead()`. In `loadWAL`, we can assume that the right number has been set and directly use it. We don't need to update `DefaultOptions()` in db.go since we do the validation in the Head.